source/common/common/trie.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+namespace {
+// Because vector is efficient when removing elements from the back but paths
+// generally have the most significant components at the front, it's useful to
+// reverse the components vector before processing
+template <class T> inline void reverseComponentVector(std::vector<T>& vec) {
```

Comment:
Please split into trie.h and trie.cc. We shouldn't be dropping implementation detail helpers into unrelated code.